### PR TITLE
Consumes migration for PFSimParticleProducer in RecoParticleFlow/PFSimProducer 

### PR DESCRIPTION
**Pull request by schoef**
Consumes migration for PFSimParticleProducer in RecoParticleFlow/PFSimProducer

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
Shouldn't the EDProducer inheritance in PFSimParticleProducer.h be fixed for thread safety? The static analyzer complains about it.

`line 44: class PFSimParticleProducer : public edm::EDProducer {`

**Comment by schoef** [comment]
@cvuosalo Is this the fix? 
`class PFSimParticleProducer : public edm::stream::EDProducer<>` {

**Comment by cvuosalo** [comment]
@schoef: Yes, I think that will work.

**Comment by cvuosalo** [comment]
@schoef: On second thought, ::stream won't won't be reliable if there is non-thread-safe code somewhere in this producer. When you make the change, please test in multi-threaded mode to check for problems.  To do so, you can append the following to your config file:

`echo -e "process.options.numberOfStreams = cms.untracked.uint32(0)\n process.options.numberOfThreads = cms.untracked.uint32(4)\n" >> configFile`

**Comment by schoef** [comment]
@cvuosalo Indeed, that test fails.
Could you let me know what else needs to change?
Thank you.

**Comment by lgray** [comment]
@schoef This code will need to be (completely) rewritten to be even thread friendly. Recommend that for now you use ```edm::one``` instead of ```edm::stream```.

**Comment by schoef** [comment]
@lgray OK, afaik this producer isn't currently used anywhere except in JME studies. That's why I'm making the small updates.
I've changed to edm::one

**Comment by lgray** [comment]
@cmsbuild please test

(Sent from my Nexus 6)

**Comment by schoef** [comment]
OK, maybe too fast ... the producer doesn't start, i.e. cmsRun hangs. 
I don't know why this is.
Tests are successful since this code is not used.

It does work if I revert to inheriting from 'EDProducer'.
I'd prefer to revert since otherwise I need to maintain. Is this possible?

**Comment by lgray** [comment]
Oh, this has the FastSim event management objects in it.
These do no play nicely with the threaded framework.
Better to leave as legacy module, just having migration to consumes is already good.

This will need to be totally rewritten at some point to bring it into the future.

**Comment by schoef** [comment]
@lgray 
OK, done.

**Comment by lgray** [comment]
@schoef then all you need to do is:
```
git reset --hard ae8fcae
git push my-cmssw CMSSW_7_6_X_devel -f
```
This will reset the pull request to the old commit and restore the signoffs (without need for testing again)

**Comment by cvuosalo** [comment]
@cmsbuild please notice reset

**Comment by cvuosalo** [comment]
@Dr15Jones: Could you please take a look here? There is a legacy EDProducer, but when it is made one::EDProducer<>, cmsRun hangs. How can we migrate this code to multi-threading without a major re-write?

**Comment by Dr15Jones** [comment]
Could you run a job with the change under gdb and when it locks up do a 
```thread apply all backtrace```?

**Comment by Dr15Jones** [comment]
On a quick inspection, it looks like the module could be ```stream::``` since it appears attempts were made to make FastSim parts thread safe:
http://cmslxr.fnal.gov/source/FastSimulation/Particle/interface/ParticleTable.h?v=CMSSW_8_0_X_2015-11-23-2300#0039

the static analyzer didn't find any statics associated with the producer (although the static analyzer has been known to miss items in the past).

**Comment by schoef** [comment]
@Dr15Jones Any of these action items for me? If so I need more detailed instructions.

**Comment by Dr15Jones** [comment]
1. run the job which 'hangs' under gdb: ```gdb --args cmsRun <name of config>``` then ```gdb>run```.
2. once job hangs, do a ```<control>c```
3. in the gdb prompt do ```gdb>thread apply all backtrace```
4. past the output into a comment to this pull request.

**Comment by schoef** [comment]
I locally changed back to `edm::one::EDProducer<>` 
Hope this helps:
```
gdb --args cmsRun test.py
Name collision between vt102 vt102
Name collision between vt220 vt220
Name collision between vt100-nav vt100-nav
Name collision between vt100 vt100
Name collision between vt52 vt52
Name collision between ansi ansi
Name collision between linux linux
Name collision between dumb dumb
GNU gdb (GDB) 7.9.1
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from cmsRun...(no debugging symbols found)...done.
(gdb) run
Starting program: /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/bin/slc6_amd64_gcc493/cmsRun test.py
warning: File "/cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6.0.20-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
	add-auto-load-safe-path /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6.0.20-gdb.py
line to your configuration file "/afs/hephy.at/user/r/rschoefbeck/.gdbinit".
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/afs/hephy.at/user/r/rschoefbeck/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
# Conditions read from  CMS_CONDITIONS  via FrontierProd 
Internal FlatRandomEGun is initialzed
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 at 02-Dec-2015 09:55:04.515 CET
%MSG-w InvalidInputTag:  CSCHaloDataProducer:CSCHaloData  02-Dec-2015 09:55:48 CET Run: 1 Event: 1
The HLT results do not appear to be in the event. The beam halo HLT trigger decision will not be used in the halo identification
%MSG
^C[New Thread 0x7fffd4922700 (LWP 8983)]
[New Thread 0x7fffe0e7f700 (LWP 8577)]

Program received signal SIGINT, Interrupt.
0x00007fffd74af8ea in std::_Rb_tree<HepPDT::ParticleID, std::pair<HepPDT::ParticleID const, HepPDT::ParticleData>, std::_Select1st<std::pair<HepPDT::ParticleID const, HepPDT::ParticleData> >, HepPDT::ParticleDataTableComparison, std::allocator<std::pair<HepPDT::ParticleID const, HepPDT::ParticleData> > >::find(HepPDT::ParticleID const&) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/external/slc6_amd64_gcc493/lib/libHepPDT.so
```
Otherwise, I see no problem with the current code.

**Comment by Dr15Jones** [comment]
@schoef I think you may have forgetten to do ```thread apply all backtrace``` after doing the control C. Also are you sure the job was actually stuck by that time? From the function that was printed out it looks like the job was doing work on a map. Also, I don't see the printout stating the job is using more than one thread.

**Comment by schoef** [comment]
Well, I let it hang for 10', it usually completes within 1'.
It hangs without multithreading.
```
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 at 02-Dec-2015 14:00:33.449 CET
%MSG-w InvalidInputTag:  CSCHaloDataProducer:CSCHaloData  02-Dec-2015 14:01:13 CET Run: 1 Event: 1
The HLT results do not appear to be in the event. The beam halo HLT trigger decision will not be used in the halo identification
%MSG
^C[New Thread 0x7fffd4922700 (LWP 23213)]
[New Thread 0x7fffe0e7f700 (LWP 23173)]

Program received signal SIGINT, Interrupt.
0x00007fffd74acc10 in HepPDT::ParticleID::abspid() const () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/external/slc6_amd64_gcc493/lib/libHepPDT.so
(gdb) thread apply all backtrace

Thread 3 (Thread 0x7fffe0e7f700 (LWP 23173)):
#0  0x000000306960e7dd in read () from /lib64/libpthread.so.0
#1  0x00007fffebd4585a in full_read.constprop () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreServices.so
#2  0x00007fffebd459c9 in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreServices.so
#3  0x00007ffff6962ac0 in execute_native_thread_routine () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6
#4  0x0000003069607a51 in start_thread () from /lib64/libpthread.so.0
#5  0x00000030692e896d in clone () from /lib64/libc.so.6

Thread 2 (Thread 0x7fffd4922700 (LWP 23213)):
#0  0x000000306960b63c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff695d758 in __gthread_cond_wait(pthread_cond_t*, pthread_mutex_t*) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6
#2  0x00007ffff695d7f6 in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6
#3  0x00007fffd4c8f533 in OscarMTMasterThread::OscarMTMasterThread(edm::ParameterSet const&)::{lambda()#1}::operator()() const ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/biglib/slc6_amd64_gcc493/pluginSimulation.so
#4  0x00007ffff6962ac0 in execute_native_thread_routine () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6
#5  0x0000003069607a51 in start_thread () from /lib64/libpthread.so.0
#6  0x00000030692e896d in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7ffff6685a40 (LWP 23130)):
#0  0x00007fffd74acc10 in HepPDT::ParticleID::abspid() const () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/external/slc6_amd64_gcc493/lib/libHepPDT.so
#1  0x00007fffd74af8ea in std::_Rb_tree<HepPDT::ParticleID, std::pair<HepPDT::ParticleID const, HepPDT::ParticleData>, std::_Select1st<std::pair<HepPDT::ParticleID const, HepPDT::ParticleData> >, HepPDT::ParticleDataTableComparison, std::allocator<std::pair<HepPDT::ParticleID const, HepPDT::ParticleData> > >::find(HepPDT::ParticleID const&) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/external/slc6_amd64_gcc493/lib/libHepPDT.so
#2  0x00007fffd74af751 in HepPDT::ParticleDataTable::particle(HepPDT::ParticleID) const () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/external/slc6_amd64_gcc493/lib/libHepPDT.so
#3  0x00007fffc420f057 in RawParticle::setID(int) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFastSimulationParticle.so
#4  0x00007fffbb24c74c in FBaseSimEvent::fill(std::vector<SimTrack, std::allocator<SimTrack> > const&, std::vector<SimVertex, std::allocator<SimVertex> > const&) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFastSimulationEvent.so
#5  0x00007fffbb24e8cd in FSimEvent::fill(std::vector<SimTrack, std::allocator<SimTrack> > const&, std::vector<SimVertex, std::allocator<SimVertex> > const&) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFastSimulationEvent.so
#6  0x00007fffb92b110f in PFSimParticleProducer::produce(edm::Event&, edm::EventSetup const&) ()
   from /afs/hephy.at/work/r/rschoefbeck/CMS/CMSSW_7_6_1/lib/slc6_amd64_gcc493/pluginRecoParticleFlowPFProducerSimPlugins.so
#7  0x00007ffff7d5a668 in edm::one::EDProducerBase::doEvent(edm::EventPrincipal&, edm::EventSetup const&, edm::ActivityRegistry*, edm::ModuleCallingContext const*) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#8  0x00007ffff7d5318f in edm::WorkerT<edm::one::EDProducerBase>::implDo(edm::EventPrincipal&, edm::EventSetup const&, edm::ModuleCallingContext const*) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#9  0x00007ffff7c8d0bc in decltype ({parm#1}()) edm::convertException::wrap<bool edm::Worker::doWork<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}>(bool edm::Worker::doWork<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#10 0x00007ffff7c8d374 in bool edm::Worker::doWork<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#11 0x00007ffff7c9343e in decltype ({parm#1}()) edm::convertException::wrap<void edm::Path::processOneOccurrence<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, edm::StreamID const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}>(void edm::Path::processOneOccurrence<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, edm::StreamID const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#12 0x00007ffff7c93702 in void edm::Path::processOneOccurrence<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, edm::StreamID const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#13 0x00007ffff7c93c0c in void edm::StreamSchedule::processOneEvent<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, bool)::{lambda()#1}::operator()() const () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#14 0x00007ffff7c93deb in decltype ({parm#1}()) edm::convertException::wrap<void edm::StreamSchedule::processOneEvent<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, bool)::{lambda()#1}>(void edm::StreamSchedule::processOneEvent<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, bool)::{lambda()#1}) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#15 0x00007ffff7c94048 in void edm::StreamSchedule::processOneEvent<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal&, edm::EventSetup const&, bool) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#16 0x00007ffff7c88795 in edm::EventProcessor::processEvent(unsigned int) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
---Type <return> to continue, or q <return> to quit---
#17 0x00007ffff7c88fc2 in edm::EventProcessor::processEventsForStreamAsync(unsigned int, std::atomic<bool>*) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#18 0x00007ffff7c98644 in edm::StreamProcessingTask::execute() () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#19 0x00007ffff6a26bec in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7ffff55e7d00, parent=..., child=<optimized out>)
    at ../../src/tbb/custom_scheduler.h:474
#20 0x00007ffff7c88cce in edm::EventProcessor::readAndProcessEvent() () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#21 0x00007ffff7c654cd in statemachine::HandleEvent::readAndProcessEvent() () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#22 0x00007ffff7c67016 in statemachine::HandleEvent::HandleEvent(boost::statechart::state<statemachine::HandleEvent, statemachine::HandleLumis, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::my_context) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#23 0x00007ffff7c6fac7 in boost::statechart::state<statemachine::HandleEvent, statemachine::HandleLumis, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::deep_construct(boost::intrusive_ptr<statemachine::HandleLumis> const&, boost::statechart::state_machine<statemachine::Machine, statemachine::Starting, std::allocator<void>, boost::statechart::null_exception_translator>&) ()
   from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#24 0x00007ffff7c6fcbc in boost::statechart::simple_state<statemachine::FirstLumi, statemachine::HandleLumis, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::react_impl(boost::statechart::event_base const&, void const*) () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#25 0x00007ffff7c7fa64 in edm::EventProcessor::runToCompletion() () from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_7_6_1/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#26 0x000000000040d048 in main::{lambda()#1}::operator()() const ()
#27 0x000000000040b4a6 in main ()
```

**Comment by Dr15Jones** [comment]
I think I found the problem. Your ParticleTable is not being initialized becuase your module's ```beginRun``` is not being called. But the compiler would have caught that as long as you had ```override``` declared on the declaration of ```beginRun```. When you changed to using a ```edm::one::EDProducer<>``` did you also remove the ```override```?

**Comment by Dr15Jones** [comment]
A ```edm::one::EDProducer``` can state it needs to know about Run transitions but you must explicitly declare that as part of its template arguments. The documentation is
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface#edm_one_WatchRuns

But it just boils down to inheriting from ```edm::one::EDProducer<edm::one::WatchRuns>```.

**Comment by Dr15Jones** [comment]
An even better solution is just to move the work done in ```beginRun``` to the very beginning of ```produce```.

**Comment by Dr15Jones** [comment]
The reason this happened is FBaseSimEvent doesn't initialize its member data ```pdt``` to ```nullptr``` in its constructor. Therefore it is set to some random value and then when you call a method on that pointer you get random results.

**Comment by schoef** [comment]
OK, thanks for the diagnosis.
I moved the code from beginRun to produce and it seems to work.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFSimProducer/plugins/PFSimParticleProducer.h [hunk 1]
```diff
@@ -44,7 +50,7 @@ class PFSimParticleProducer : public edm::EDProducer {
   
   virtual void produce(edm::Event&, const edm::EventSetup&) override;
 
-  virtual void beginRun(const edm::Run& r, const edm::EventSetup & c) override;
+  virtual void beginRun(const edm::Run& r, const edm::EventSetup & c);
```
Comment (cvuosalo) [inline RecoParticleFlow/PFSimProducer/plugins/PFSimParticleProducer.h]: `override` should not be deleted. If there is a reason to delete it, you need to add a comment explaining why this virtual function that overrides the original `beginRun` is not marked with `override`.

**Comment by cvuosalo** [comment]
@schoef: What happened to changing the inheritance from the thread-unsafe EDProducer to `edm::one::EDProducer<edm::one::WatchRuns>`? Chris seems to say that it should work.

**Comment by Dr15Jones** [comment]
Looks to me like the beginRun method should just be removed and the inheritance stay what it is now since the work that was supposed to be done in beginRun is now done in produce.

**Comment by schoef** [comment]
WIth the current inheritance it compiled and runs fine. I removed beginRun. 
The other inheritances didn't compile, smth with the override and beginRun (which I couldn't remove). I didn't spend much time on the problem though.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@cmsbuild please test
Try again in case error is transient

File: RecoParticleFlow/PFSimProducer/plugins/PFSimParticleProducer.h [hunk 2]
```diff
@@ -44,7 +50,7 @@ class PFSimParticleProducer : public edm::EDProducer {
   
   virtual void produce(edm::Event&, const edm::EventSetup&) override;
 
-  virtual void beginRun(const edm::Run& r, const edm::EventSetup & c) override;
+//virtual void beginRun(edm::Run const & r, const edm::EventSetup & c) override;
```
Comment (Dr15Jones) [inline RecoParticleFlow/PFSimProducer/plugins/PFSimParticleProducer.h]: please just remove this line

**Comment by Dr15Jones** [comment]
I believe this can safely be made into a ```stream::``` module.

**Comment by cvuosalo** [comment]
@schoef: Chris says `edm::stream::EDProducer<>` should work for replacing the thread-unsafe EDProducer. Does it?

**Comment by schoef** [comment]
Yes, that works.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #12496 7684cc2d05dc84451a951db7b403a7d9af744a87

Migrating PFSimParticleProducer to support thread safety. There should be no change in monitored quantities.

The code changes are satisfactory, and Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12496/10261/summary.html) against baseline CMSSW_7_6_X_2015-12-10-2300 show no significant differences, as expected.

**Comment by cvuosalo** [comment]
@schoef: Please make an 80X version of this PR. Thanks.

**Comment by davidlange6** [comment]
+1

**Pull request merged**